app/controllers/api/v1/courses_controller.rb

```diff
@@ -0,0 +1,120 @@
+include ActionController::Flash
+
+class Api::V1::CoursesController < ApplicationController
+  before_action :set_course, only: %i[ show update destroy ]
+
+  # GET /courses
+  # List all the courses
+  def index
+    @courses = Course.all
+
+    render json: @courses
+  end
+
+  # GET /courses/1
+  # Get a course
+  def show
+    render json: @course
+  end
+
+  # POST /courses
+  # Create a course
+  def create
+    @course = Course.new(course_params)
+
+    if @course.save
+      render json: @course, status: :created
+    else
+      render json: @course.errors, status: :unprocessable_entity
+    end
+  end
+
+  # PATCH/PUT /courses/1
+  # Update a course
+  def update
+    if @course.update(course_params)
+      render json: @course
+    else
+      render json: @course.errors, status: :unprocessable_entity
+    end
+  end
+
+  # DELETE /courses/1
+  # Delete a course
+  def destroy
+    @course.destroy
+    render json: { message: "Course with id #{params[:id]}, deleted" }, status: :ok
+  end
+
+  # Adds a Teaching Assistant to the course
+  def add_ta
```

Comment:
By moving the logic to the model, the controller is responsible only for handling the request and sending out the appropriate response. 

Possible Refactor:
In the `Course` model:
```ruby
class Course < ApplicationRecord
  def add_ta(user)
    if user.nil?
      return { success: false, message: "The user #{user.id} does not exist" }
    elsif TaMapping.exists?(ta_id: user.id, course_id: id)
      return { success: false, message: "The user #{user.id} is already a TA for this course." }
    else
      ta_mapping = TaMapping.create(ta_id: user.id, course_id: id)
      user.update(role: Role::TEACHING_ASSISTANT)

      if ta_mapping.save
        return { success: true, data: ta_mapping.slice(:course_id, :ta_id) }
      else
        return { success: false, message: ta_mapping.errors }
      end
    end
  end
end
```

In the controller:
```ruby
def add_ta
  user = User.find_by(id: params[:ta_id])
  result = @course.add_ta(user)

  if result[:success]
    render json: result[:data], status: :created
  else
    render json: { status: "error", message: result[:message] }, status: :bad_request
  end
end
```

To streamline the code, consider adding "add_ta" to the list of actions in the `before_action: set_course` method instead of creating a new local variable for the course.

Reply:
Thanks for the suggestion; I will separate the logic from the controller and add add_ta to the list of actions.